clientd3d/preferences.c

```diff
@@ -0,0 +1,1684 @@
+    if (strcmp(TCCompare, TCMapzoomin) == 0 && TCCompare != TCMapzoomin) return true;
+    if (strcmp(TCCompare, TCMapzoomout) == 0 && TCCompare != TCMapzoomout) return true;
+    return false;
+}
+
+static void CheckforDuplicateBindM(const char *TCCompare)
+{
+    if (CheckforDuplicateBind(TCCompare))
+    {
+        MessageBox(NULL, GetString(hInst, IDS_DUPLICATE_BINDINGS), GetString(hInst, IDS_WARNING_TITLE), MB_OK | MB_ICONEXCLAMATION);
+    }
+}
+
+// Function to handle key assignment
+static void AssignKey(HWND hDlg, TCHAR* TCButtonstring, int nID) {
+    // Show the Assign Key dialog
+    if (DialogBox(GetModuleHandle(NULL), MAKEINTRESOURCE(IDD_ASSIGNKEY), hDlg, AssignKeyDlgProc) == IDOK) {
+        strcpy_s(TCButtonstring, MAX_KEYVALUELEN, TCNewkey);
+        SetDlgItemText(hDlg, nID, TCButtonstring);
+        CheckforDuplicateBindM(TCButtonstring);
+    }
+}
+
+int iModifier;
+HHOOK hHook = NULL;
+
+// Dialog procedure for the Modifier dialog
+INT_PTR CALLBACK ModifierDlgProc(HWND hDlg, UINT message, WPARAM wParam, LPARAM lParam)
+{
+    switch (message) {
+    case WM_INITDIALOG:
+        return (INT_PTR)TRUE;
+
+    case WM_COMMAND:
+        switch (LOWORD(wParam)) {
+        case IDC_ALT:
+            iModifier = MODIFIER_ALT;
+            EndDialog(hDlg, IDOK);
+            return (INT_PTR)TRUE;
+        case IDC_CTRL:
+            iModifier = MODIFIER_CTRL;
+            EndDialog(hDlg, IDOK);
+            return (INT_PTR)TRUE;
+        case IDC_SHIFT:
+            iModifier = MODIFIER_SHIFT;
+            EndDialog(hDlg, IDOK);
+            return (INT_PTR)TRUE;
+        case IDC_ANY:
+            iModifier = MODIFIER_ANY;
+            EndDialog(hDlg, IDOK);
+            return (INT_PTR)TRUE;
+        case IDC_NONE:
+            iModifier = MODIFIER_NONE;
+            EndDialog(hDlg, IDOK);
+            return (INT_PTR)TRUE;
+        case IDCANCEL:
+            EndDialog(hDlg, IDCANCEL);
+            return (INT_PTR)TRUE;
+        }
+        break;
+    }
+    return (INT_PTR)FALSE;
+}
+
+// Function to handle key modification
+static void ModifyKey(HWND hDlg, TCHAR* TCString, int nID)
+{
+    // Show the Modifier dialog
+    if (DialogBox(GetModuleHandle(NULL), MAKEINTRESOURCE(IDD_MODIFIER), hDlg, ModifierDlgProc) == IDOK)
+    {
+        strncat(TCString, "+", MAX_KEYVALUELEN - strlen(TCString) - 1);
+
+        const char* modifierStr =
+            (iModifier == MODIFIER_ALT) ? "alt" :
+            (iModifier == MODIFIER_CTRL) ? "ctrl" :
+            (iModifier == MODIFIER_SHIFT) ? "shift" :
+            (iModifier == MODIFIER_ANY) ? "any" : "none";
+
+        strncat(TCString, modifierStr, MAX_KEYVALUELEN - strlen(TCString) - 1);
+        SetDlgItemText(hDlg, nID, TCString);
+    }
+}
+
+static void RestoreDefaults(HWND hDlg)
+{
+    wchar_t szBuffer[256];
+    char szBufferA[256];
+
+    MultiByteToWideChar(CP_ACP, 0, DEF_INVERT, -1, szBuffer, 256);
+    WideCharToMultiByte(CP_ACP, 0, szBuffer, -1, szBufferA, 256, NULL, NULL);
```

Comment:
Do you know what's going on here with this multibyte stuff?  It seems unnecessary.